STARGenerateGenome <- function(rna_params){
  ##STARGenomeGenerate
  ##Description:
  ##Takes the parameters and organizes a function from STAR to generate a genome 
  ##Arguments:
  ##rna_params : parameter values stored in a list 
  ##Returns:
  ##None
  #Set the working directory to the STAR directory
  setwd(rna_params[2])
  #Create string variables
  fasta <- ""
  loadFile <- ""
  command <- ""
  dir.create("GenomeDir")
  ##Loops through the files and forms a command to STAR  
  genomeFiles <- dir(rna_params[4], pattern = ".fa")
  for (file in genomeFiles){
    loadFile <- paste(rna_params[4],"/",file, sep = "", collapse = NULL)
    fasta <- paste(fasta, loadFile, sep = " ", collapse = NULL)
  }
  #Creating the command
  command <- paste("STAR --runThreadN ", rna_params[1], " --runMode genomeGenerate --genomeDir GenomeDir --genomeFastaFiles ", fasta, " --sjdbGTFfile ", rna_params[3], " --sjdbOverhang ", as.numeric(rna_params[5]) - 1, sep = "", collapse = NULL )
  #Running the command
  print(command)
  system(command)
}

STARAlign <- function(rna_params){
  ##STAR AlignReads
  ##Description:
  ##Takes the parameters and organizes a STAR function to align the RNASEQ files to the genome
  ##Arguments
  ##rna_params : parameter values stored in a list 
  ##Returns
  ##None                                                                                                  
  setwd(rna_params[2])
  
  #Reads the files from the fastq file directory and puts them into a list
  alignFiles <- dir(rna_params[6], pattern = ".fastq")
  
  #Create the directories to put the files into after they are generated by STAR
  dir.create("Results")
  dir.create("Results/AlignedFiles")
  dir.create("Results/STAR_metadata_logout")
  dir.create("Results/STAR_metadata_finallog")
  dir.create("Results/STAR_metadata_progresslog")
  dir.create("Results/STAR_Transcriptome")
  dir.create("Results/STAR_metadata_SJ")
  dir.create("Results/STAR_metadata_ReadsPerGene")
  
  #Loops through the files and runs STAR to get a series of files (SAM format align file and all of its log files)
  if (rna_params[9] == "Y"){
    filenames <- list.files(path = rna_params[6], full.names = FALSE, recursive = FALSE)
    for (index in 1:length(filenames)){
      if (index %% 2 == 0){
        next
      }
      else{
        inputfile <- paste(rna_params[6], "/", filenames[index], " ", rna_params[6], "/", filenames[index + 1], sep = "", collapse = NULL)
        command <- paste("STAR --runMode alignReads --runThreadN ", rna_params[1], " --quantMode TranscriptomeSAM GeneCounts --outSAMtype SAM  --outSAMunmapped Within --outFilterScoreMinOverLread 0.3 --outFilterMatchNminOverLread 0.3 --genomeDir GenomeDir --readFilesIn ", inputfile, sep = "", collapse = NULL)
        print(command)
        system(command)
        statement <- paste("The ", filenames[index], " has been aligned")
        print(statement)
        #Renames and moves the files to their designated folders specified in the parameter file
        file.rename("Aligned.out.sam", paste("Results/AlignedFiles/", filenames[index], ".sam", sep = "", collapse = NULL))
        file.rename("Log.out", paste("Results/STAR_metadata_logout/", filenames[index], "logout.txt", sep = "", collapse = NULL))
        file.rename("Log.final.out", paste("Results/STAR_metadata_finallog/", filenames[index], "finlogout.txt", sep = "", collapse = NULL))
        file.rename("Log.progress.out", paste("Results/STAR_metadata_progresslog/", filenames[index], "logprogress.txt", sep = "", collapse = NULL))
        file.rename("Aligned.toTranscriptome.out.bam", paste("Results/STAR_Transcriptome/", filenames[index], "transcriptome.bam", sep = "", collapse = NULL))
        file.rename("SJ.out.tab", paste("Results/STAR_metadata_SJ/", filenames[index], "SJout.tab", sep = "", collapse = NULL))
        file.rename("ReadsPerGene.out.tab", paste("Results/STAR_metadata_ReadsPerGene/", filenames[index], "ReadsPerGene", sep = "", collapse = NULL))
      }
    } 
  }
  
  else if (rna_params[9] == "N"){
    count <- 0
    for (file in alignFiles){
      command <- paste("STAR --runMode alignReads --runThreadN ", rna_params[1], " --quantMode TranscriptomeSAM GeneCounts --outSAMtype SAM --outSAMunmapped Within --genomeDir GenomeDir --readFilesIn ", rna_params[6], "/", file, sep = "", collapse = NULL)
      print(command)
      system(command)
      statement <- paste("The ", file, " has been aligned")
      print(statement)
      #Renames and moves the files to their designated folders specified in the parameter file
      file.rename("Aligned.out.sam", paste("Results/AlignedFiles/", file, ".sam", sep = "", collapse = NULL))
      file.rename("Log.out", paste("Results/STAR_metadata_logout/", file, "logout.txt", sep = "", collapse = NULL))
      file.rename("Log.final.out", paste("Results/STAR_metadata_finallog/", file, "finlogout.txt", sep = "", collapse = NULL))
      file.rename("Log.progress.out", paste("Results/STAR_metadata_progresslog/", file, "logprogress.txt", sep = "", collapse = NULL))
      file.rename("Aligned.toTranscriptome.out.bam", paste("Results/STAR_Transcriptome/", file, "transcriptome.bam", sep = "", collapse = NULL))
      file.rename("SJ.out.tab", paste("Results/STAR_metadata_SJ/", file, "SJout.tab", sep = "", collapse = NULL))
      file.rename("ReadsPerGene.out.tab", paste("Results/STAR_metadata_ReadsPerGene/", file, "ReadsPerGene", sep = "", collapse = NULL))
    }
  }
}

gunzip_check <- function(rna_params){
  ##Gunzip check  
  ##Description:
  ##Takes the gzipped files from a folder and unzips them. 
  ##Arguments:      
  ##rna_params: parameter values stored in a list 
  ##Returns:
  ##None
  setwd(rna_params[2])
  if (rna_params[7] == "Y"){'/media/nick/LinuxData/RNAseq/LUCA RNASEQ/luca_analysis_final/Results/Enrichment/'
    gzFiles <- dir(rna_params[6], pattern = ".gz")
    for (file in gzFiles){
      command <- paste("gunzip ", rna_params[6], "/", file, sep = "", collapse = NULL)
      system(command)
    }
  }
  else if (rna_params[7] == "N"){
    print("No zipped files was selected, continuing")
  }
}         
